src/boot/bootctl.c

```diff
@@ -2555,6 +2556,334 @@ static int verb_reboot_to_firmware(int argc, char *argv[], void *userdata) {
         }
 }
 
+/* This only frees some fields. See boot_entry_auto */
+static void boot_entry_auto_done(BootEntry* entry) {
+        free(entry->title);
+        free(entry->version);
+        free(entry->sort_key);
+        free(entry->machine_id);
+
+        if (entry->options) {
+                for (size_t i = 0; i < SIZE_MAX; i++)
+                {
+                        if (!entry->options[i])
+                                break;
+
+                        free(entry->options[i]);
+                }
+        }
+        free(entry->options);
+}
+
+static char* boot_entry_to_string(const BootEntry* entry){
+        char* s = strjoin("linux    ", entry->kernel, "\n");
+
+        if (entry->title) {
+                s = strpush(s, "title    ", entry->title, "\n");
+                if (!s) {
+                        free(s);
+                        return NULL;
+                }
+        }
+
```

Comment:
Each append here copies the string, so it scales as the square of the number of strings added. We generally don't have too many strings, but this is still pretty ugly. `open_memstream_unlocked()` or a normal file should be used instead.

Reply:
Just changed it to use a caller-provided stream.